### PR TITLE
feat(masthead): overflow feature

### PR DESCRIPTION
**Pull request by ariellalgilmore**
### Related Ticket(s)

#3875 

### Description

change mastheadTopNav to include the overflow feature if menuItems go beyond the container width arrows will be added for navigation

### Changelog

**New**

- add left and right caret with functions for onClick and resize effect 
- added styles for the left and right carets

Screen Shot 2020-09-25 at 7 44 32 AM

### CODE REVIEW

File: packages/styles/scss/components/masthead/_masthead.scss [hunk 1]
```diff
@@ -290,8 +296,76 @@ $search-transition-timing: 95ms;
     &:active {
       border-color: $focus;
       background-color: $hover-ui;
+      display: block;
+      &::before,
+      &::after {
+        display: none;
+      }
+    }
+    &-left-caret,
```
Comment (asudoh) [inline packages/styles/scss/components/masthead/_masthead.scss:305]: We are going away from BEM concatenation (`&-foo`) for better reuse of style rules between React and Web Components.

File: packages/react/src/components/Masthead/MastheadTopNav.js [hunk 2]
```diff
@@ -23,6 +28,65 @@ const { prefix } = settings;
...
+  const [overflow, setOverflow] = useState(false);
+  const [showLeftCaret, setShowLeftCaret] = useState(false);
+
+  useEffect(() => {
+    onResize();
+    root.addEventListener('resize', onResize);
+    return () => {
+      root.removeEventListener('resize', onResize);
+    };
+  });
```
Comment (asudoh) [inline packages/react/src/components/Masthead/MastheadTopNav.js:41]: Given we seem to want to track the size of `.bx--header__nav-container`, good to track its size directly via [`ResizeObserveer`](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver).

**Review by asudoh — COMMENTED** [review]
Nice UI, a couple of comments - Thanks @ariellalgilmore!

File: packages/styles/scss/components/masthead/_masthead.scss (hunk 1 above)
Comment (ariellalgilmore) [reply packages/styles/scss/components/masthead/_masthead.scss]: fixed! Thanks @asudoh!

**Comment by jeffchew** [comment]
@ariellalgilmore I was testing it out and didn't seem to work. Clicking the arrows didn't do anything.

File: packages/react/src/components/Masthead/MastheadTopNav.js [hunk 3]
```diff
@@ -23,7 +27,68 @@ const { prefix } = settings;
...
+      }
+    } else {
+      setOverflow(false);
+      setShowLeftCaret(false);
+      setShowRightCaret(false);
+    }
+  });
+
+  useEffect(() => {
+    resizeObserver.observe(topNavProps.topNavRef.current);
```
Comment (asudoh) [inline packages/react/src/components/Masthead/MastheadTopNav.js:52]: Any reason the DOM ref has to come from `<Masthead>` instead of `<div class="bx--header__nav-container">` that this component renders?

File: packages/react/src/components/Masthead/MastheadTopNav.js [hunk 4]
```diff
@@ -23,7 +27,68 @@ const { prefix } = settings;
...
+  let containerWidth = 0;
+  const resizeObserver = new ResizeObserver(() => {
+    containerWidth =
+      calculateTotalWidth(['bx--header__search']) -
+      calculateTotalWidth(['bx--masthead__search']) -
+      calculateTotalWidth(['bx--header__name']);
+    const totalNavWidth = calculateTotalWidth(['bx--header__nav']);
+    if (totalNavWidth > containerWidth) {
+      setOverflow(true);
+      const offset = document.querySelector('.bx--header__nav-container')
```
Comment (asudoh) [inline packages/react/src/components/Masthead/MastheadTopNav.js:39]: Let's grab this DOM ref via React `ref`.

File: packages/react/src/components/Masthead/MastheadTopNav.js [hunk 5]
```diff
@@ -23,7 +27,68 @@ const { prefix } = settings;
...
+    resizeObserver.observe(topNavProps.topNavRef.current);
+  }, [topNavProps.topNavRef, resizeObserver]);
+
+  /**
+   * Scroll navigation left
+   */
+  function paginateLeft() {
+    document.querySelector(
+      '.bx--header__nav-container'
+    ).scrollLeft -= containerWidth;
```
Comment (asudoh) [inline packages/react/src/components/Masthead/MastheadTopNav.js:61]: Let's grab this DOM ref via React `ref`.

File: packages/react/src/components/Masthead/MastheadTopNav.js [hunk 6]
```diff
@@ -23,7 +27,68 @@ const { prefix } = settings;
...
+    if (showLeftCaret) {
+      document.querySelector('.bx--header__nav-container').scrollLeft -= 40;
```
Comment (asudoh) [inline packages/react/src/components/Masthead/MastheadTopNav.js:63]: Let's grab this DOM ref via React `ref`.

File: packages/react/src/components/Masthead/MastheadTopNav.js [hunk 7]
```diff
@@ -23,7 +27,68 @@ const { prefix } = settings;
...
+    }
+    setShowRightCaret(true);
+    if (document.querySelector('.bx--header__nav-container').scrollLeft <= 0) {
```
Comment (asudoh) [inline packages/react/src/components/Masthead/MastheadTopNav.js:66]: Let's grab this DOM ref via React `ref`.

File: packages/react/src/components/Masthead/MastheadTopNav.js [hunk 8]
```diff
@@ -23,7 +27,68 @@ const { prefix } = settings;
...
+      setShowLeftCaret(false);
+    }
+  }
+  /**
+   * Scroll navigation right
+   */
+  function paginateRight() {
+    document.querySelector(
+      '.bx--header__nav-container'
+    ).scrollLeft += containerWidth;
```
Comment (asudoh) [inline packages/react/src/components/Masthead/MastheadTopNav.js:76]: Let's grab this DOM ref via React `ref`.

File: packages/react/src/components/Masthead/MastheadTopNav.js [hunk 9]
```diff
@@ -23,7 +27,68 @@ const { prefix } = settings;
...
+    setShowLeftCaret(true);
+    const totalNavWidth = calculateTotalWidth(['bx--header__nav']);
+    if (
+      document.querySelector('.bx--header__nav-container').scrollLeft +
```
Comment (asudoh) [inline packages/react/src/components/Masthead/MastheadTopNav.js:80]: Let's grab this DOM ref via React `ref`.

**Review by asudoh — COMMENTED** [review]
Looks that you are getting there 🎉 - Thanks @ariellalgilmore!

File: packages/react/src/components/Masthead/MastheadTopNav.js (hunk 3 above)
Comment (ariellalgilmore) [reply packages/react/src/components/Masthead/MastheadTopNav.js:52]: The DOM ref has to come from `<Masthead>` because if the user starts in wider screen and then shrinks the size where the overflow should turn on then only the `<Masthead>` will change in size. An overflow style causes `<div class="bx--header__nav-container">` to stay the same size even if the user shrinks the browser, which will then not trigger the resizeObserver.

**Review by asudoh — APPROVED** [review]
The latest code looks 💯🎉  - The only question is the visual change of mobile search in its open state (not sure if it's intentional or not). Thank you so much @ariellalgilmore!

**Review by jeffchew — CHANGES_REQUESTED** [review]
Hey @ariellalgilmore , can you take a look at what's going on in mobile?

Screenshot 2020-10-02 at 7 36 15 AM

Also, for the custom data so we can test the overflow, perhaps instead of this data, let's have multiple menu items with either lorem ipsum text or just `Link 1`, `Link 2`, etc:

Screenshot 2020-10-02 at 7 36 00 AM

**Comment by jeffchew** [comment]
@ariellalgilmore can you rebase with latest as well? I think it doesn't include the L1 out of feature flag so want to see if the overflow is working there too.

**Comment by ariellalgilmore** [comment]
@jeffchew good call!

**Comment by ariellalgilmore** [comment]
closing due to many merge conflicts and opening new PR
